### PR TITLE
CBG-345 - Channel Cache compaction

### PR DESCRIPTION
**Pull request by adamcfraser**
Adds compaction for channel cache.  Moves channel cache collection (set of singleChannelCache) to a new RangeSafeCollection to support concurrent iteration and append.  This also reduces cache contention for other cache operations (primarily high cached sequence tracking).

Channels are prioritized for eviction based on inactivity (connected replications) first, then by not recently used.  NRU flag is incremented during compaction, decremented on getChanges for the channel.

## Integration Tests:
- ~http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/1128/~
- ~http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/1130/~
- [x] http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/1137/
- [ ] http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/1139/

### CODE REVIEW

File: db/channel_cache.go [hunk 1]
```diff
@@ -214,831 +245,202 @@ func (c *channelCacheImpl) ReleaseLateSequenceClient(channelName string, current
...
+//	//     2. addChannelCache obtains stable sequence, seq=10
+//	//     3. addToCache (from another goroutine) receives sequence 11 in channel A, but detects that it's inactive (not in c.channelCaches)
+//	//     4. addChannelCache initializes cache with validFrom=10 and adds to c.channelCaches
+//	//  This scenario would result in sequence 11 missing from the cache.  Locking seqLock ensures that
+//	//  step 3 blocks until step 4 is complete (and so sees the channel as active)
+func (c *channelCacheImpl) addChannelCache(channelName string) *singleChannelCache {
 
-func newChannelCacheWithOptions(context *DatabaseContext, channelName string, validFrom uint64, options ChannelCacheOptions) *singleChannelCache {
-	cache := newChannelCache(context, channelName, validFrom)
+	// TODO: Lock seqLock instead of cacheLock
```
Comment (bbrks) [inline db/channel_cache.go]: Looks like it's done

```suggestion
```

File: db/channel_cache.go [hunk 2]
```diff
@@ -214,831 +245,202 @@ func (c *channelCacheImpl) ReleaseLateSequenceClient(channelName string, current
...
-		return c.lateLogs[len(c.lateLogs)-1]
-	}
-	return nil
-}
+		cacheSize = c.channelCaches.RemoveElements(evictionElements)
 
-func (c *singleChannelCache) addDocIDs(changes LogEntries) {
-	for _, change := range changes {
-		c.cachedDocIDs[change.DocID] = struct{}{}
+		base.Tracef(base.KeyCache, "Compact iteration complete - eviction count: %d (lwm:%d)", len(evictionElements), c.compactLowWatermark)
```
Comment (bbrks) [inline db/channel_cache.go:443]: What do you think of adding a couple of new stats to track counts of NRU Evictions and Inactive evictions?

File: db/channel_cache.go [hunk 3]
```diff
@@ -214,831 +245,202 @@ func (c *channelCacheImpl) ReleaseLateSequenceClient(channelName string, current
...
-func (c *singleChannelCache) addToCache(change *LogEntry, isRemoval bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	if c.wouldBeImmediatelyPruned(change) {
-		base.Infof(base.KeyCache, "Not adding change #%d doc %q / %q ==> channel %q, since it will be immediately pruned",
-			change.Sequence, base.UD(change.DocID), change.RevID, base.UD(c.channelName))
-		return
+	if created {
+		c.statsMap.Add(base.StatKeyChannelCacheNumChannels, 1)
```
Comment (bbrks) [inline db/channel_cache.go:308]: Do we need to decrement this stat on eviction? Or are we treating it as a "Total num of channel caches created ever" type stat?

**Review by bbrks — CHANGES_REQUESTED** [review]
Couple of questions about stats, and some small tidy up

File: db/channel_cache.go (hunk 2 above)
Comment (adamcfraser) [reply db/channel_cache.go:443]: Agree - will split that out in https://issues.couchbase.com/browse/CBG-401

File: db/channel_cache.go (hunk 3 above)
Comment (adamcfraser) [reply db/channel_cache.go:308]: Good catch - will add this.

**Comment by adamcfraser** [comment]
Updated based on PR feedback, rebased on master.

**Review by bbrks — APPROVED** [review]

**Pull request merged**